protos/flyteidl/datacatalog/datacatalog.proto

```diff
@@ -150,6 +158,39 @@ message ListDatasetsResponse {
     string next_token = 2;
 }
 
+// Get the Artifact or try to reserve a spot if the Artifact does not exist.
+message GetOrReserveArtifactRequest {
+    DatasetID dataset_id = 1;
+    string tag_name = 2;
+
+    // Minimum duration to wait from the time when the original reservation
+    // was made to consider that the original reservation has been timed-out.
+    google.protobuf.Duration timeout = 3;
```

Comment:
I don't know what's the plan for using this? How can we predict how long a task will take? I think we can have a config value for this timeout not per request... and expect the owner to keep "heartbeating" using its own unique id before the timeout expires...

Reply:
the timeout is specified by the user. It is not a prediction. Heartbeating is usually better in case you want an absolute lease. Here we are saying that this is a best effort model of trying to serialize an execution, but in the worst case, we may have concurrent executions, which is OK, as the only thing this is useful for is cost optimization. 

Heartbeating will need a change in the way propeller talks with catalog and the contract, which seems dangerous?

Reply:
So even we go with the "heartbeating" approach, we still need to specify the task execution timeout because we probably don't want a single healthy but slow task blocking other tasks from making progress. So this come down to what we want to do when encounter a timeout:
1. Kill the task execution and release the lease (task execution timeout)
2. Let the execution continue and release the lease (cache reservation timeout)

I think 1 may be an overkill for the purpose of reducing duplicated effort for populating cache since we don't necessarily need to kill the process.

Reply:
@milton0825 you are right, 1 is not the right approach, because this is not a leader election, this is purely an optimization. Haytham and I had a chat so here is an update

I agree with @EngHabu that asking the users to specify the timeout is error-prone. So instead we use a heartbeat interval like internal configuration. But then, we will need flytepropeller to keep extending the reservation

Thus the pseudo code in propeller

process 1 - gets the reservation, process -2 will get blocked

```python
before task starts:
if caching enabled:
  if cache_serialization_enabled:
        r, err = getCacheOrReservation(md=metadata, provenance-info, reservation_owner:str, 
                  reservation_timeout:duration,)
         if err { then retry with backoff }
         if r.acquired {
              goto start-task
         }
       else return not_acquired("waiting for cache")
    else: // cache serialization not enabled, but caching is
       Getcache(..) //old way

startTask:
     if caching_enabled and caching_serializatin_enabled:
           ExtendReservation(owner_id, provenance, duration)
```

`Important note.`
GetOrReserve API is invoked multple times by the same ownerID with same provenance, then only the timeout is extended.
If the timeout is expired, then GetOrReserve simply returns a new reservation (hence we need an owner id). But the semantics are crisp, there can 2 reservations that have the same provenance still in progress concurrently.
Thus if process one comes back and says, updateReservation, Datacatalog will obey it.

`Open question`
We should probably add a new API that is called Extendreservation(owner-id, timeout)

Also Flytepropeller will need a chance, to always call extendreservation and the timeout should be dynamic in  flytepropeller, this will allow controlling how often this api is invoked

Reply:
Gist:
1. Add reservation_owner and reservation_timeout as the 2 fields
2. >> Add new API ExtendReservation()
3. Document the semantics

Reply:
> GetOrReserve API is invoked multiple times by the same ownerID with same provenance, then only the timeout is extended.

What do ownerId and provenance each refer to in this context? My understanding is that we need an ID to uniquely identify a task execution that "owns" the reservation